src/cast_core/src/CollectionScanner.cpp

```diff
@@ -398,29 +398,36 @@ void CollectionScanner::run() {
             } else if (config->scanType == ScanType::kSnapshot) {
-                mongocxx::client_session session = _client->start_session({});
-                session.start_transaction(*config->transactionOptions);
-                collectionScan(config, collections, _rateLimiter, session);
-                // If a scan duration was specified, we must make the scan
-                // last at least that long.  We'll do this within any
-                // running transaction, so we keep the "long running
-                // transaction" active as long as we've been asked to.
-                // However, honor the phase's duration if specified.
-                if (config->scanDuration) {
-                    const SteadyClock::time_point now = SteadyClock::now();
-                    auto stop = started + (*config->scanDuration).value;
-                    if (config->stopPhase && *config->stopPhase < stop) {
-                        stop = *config->stopPhase;
-                    }
-                    if (stop > now) {
-                        auto sleepDuration = stop - now;
-                        auto secs = sleepDuration.count() / (1000 * 1000 * 1000);
-                        BOOST_LOG_TRIVIAL(info)
-                            << "Scanner id: " << this->_index << " sleeping " << secs;
-                        std::this_thread::sleep_for(sleepDuration);
+                try {
+                    mongocxx::client_session session = _client->start_session({});
+                    session.start_transaction(*config->transactionOptions);
+                    collectionScan(config, collections, _rateLimiter, session);
+                    // If a scan duration was specified, we must make the scan
+                    // last at least that long.  We'll do this within any
+                    // running transaction, so we keep the "long running
+                    // transaction" active as long as we've been asked to.
+                    // However, honor the phase's duration if specified.
+                    if (config->scanDuration) {
+                        const SteadyClock::time_point now = SteadyClock::now();
+                        auto stop = started + (*config->scanDuration).value;
+                        if (config->stopPhase && *config->stopPhase < stop) {
+                            stop = *config->stopPhase;
+                        }
+                        if (stop > now) {
+                            auto sleepDuration = stop - now;
+                            auto secs = sleepDuration.count() / (1000 * 1000 * 1000);
+                            BOOST_LOG_TRIVIAL(info)
+                                << "Scanner id: " << this->_index << " sleeping " << secs;
+                            std::this_thread::sleep_for(sleepDuration);
+                        }
                     }
+                    session.commit_transaction();
+                } catch (const mongocxx::operation_exception& e) {
```

Comment:
Can we narrow the exception here at all? I looked at the docs, and am not sure how much can be done http://mongocxx.org/api/current/classmongocxx_1_1operation__exception.html
Maybe transient transaction error and retryable write error labels? See https://jira.mongodb.org/browse/HELP-14706 for a discussion on error labels.

Reply:
I didn't see a way to narrow it aside from checking the labels.

I had a first version that accepted a list of labels to ignore but that seemed overkill. I'll update it to just check for transient transactions and we can see how we feel about that.

The **collectionScan** function essentially has this try / catch wrapped around internally.

Reply:
Eh, do what seems right to you. I'm asking questions, but frankly don't understand the Mongo exceptions as well as I would like.

Reply:
Scoping back to use the error label is the safest way forward for the moment. It means that the other workloads will execute exactly the same (given that they have never encountered this error anyway).

Plus it means that any new exceptions will result in failed waterfall tasks (rather than being silently eaten by the actors).